merkle-tree/concurrent/src/lib.rs

```diff
@@ -792,7 +792,12 @@ where
         leaf_index: usize,
         proof: &mut BoundedVec<[u8; 32]>,
     ) -> Result<(), ConcurrentMerkleTreeError> {
-        for changelog_entry in self.changelog.iter_from(changelog_index) {
+        // Iterate over changelog entries starting from the requested
+        // `changelog_index`.
+        //
+        // Since we are interested only in subsequent, new changelog entries,
+        // skip the first result.
+        for changelog_entry in self.changelog.iter_from(changelog_index).skip(1) {
```

Comment:
For more context - I was getting `CannotUpdateLeaf` errors in `test_100_nullify_mt` before this change. If you remove the `skip(1)`, that test is going to break.

Why? The thing is that we need to only process changelog entries **newer** than the given one. Before applying this change, I was running into edge cases, where:

* A leaf was appended, the changelog index saved.
* Some other operations were done.
* The previously appended leaf was being nullified, the saved changelog index (coming from the insertion) was being used.
* The iteration over changelog started from the insertion event. `CannotUpdateLeaf` was thrown.

You might be wondering - what if we just did `self.changelog.iter_from(changelog_index + 1)`? Wouldn't that be cleaner?

Unfortunately, that wouldn't work in case when `changelog_index` is up to date. Pointing to up-to-date `changelog_index + 1` would result in pointing to the oldest element in the whole CyclicBouldedVec, and therefore, in the full iteration over all elements.

An another alternative I was considering was skipping the first element by default in `iter_from`, but that feels wrong if we think about CyclicBoundedVec as a self-contained thing.

Reply:
True, provided the change_log index the caller provides is in sync with the root of the Merkle proof we are not interested in the changelog that was inserted when that root was created because our Merkle proof already includes that state.
This means that we have been doing unnecessary computation all the time right?
On the other hand we also cannot send changelog index + 1 to the chain because this is going to incur an iteration over the complete changelog array in case that changelog_index is the most recent changelog index  as you describe above.

Thus, skipping the first element sounds like a good solution to me.